pkg/agent/openflow/network_policy.go

```diff
@@ -710,37 +710,50 @@ func (c *client) NewDNSPacketInConjunction(id uint32) error {
 		Protocol: &protocolUDP,
 		SrcPort:  &dnsPort,
 	}
-	tcpService := v1beta2.Service{
-		Protocol: &protocolTCP,
-		SrcPort:  &dnsPort,
-	}
 	dnsPriority := priorityDNSIntercept
 	conj.serviceClause = conj.newClause(1, 2, getTableByID(conj.ruleTableID), nil)
 	conj.toClause = conj.newClause(2, 2, getTableByID(conj.ruleTableID), nil)
 	c.featureNetworkPolicy.conjMatchFlowLock.Lock()
 	defer c.featureNetworkPolicy.conjMatchFlowLock.Unlock()
 	ctxChanges := conj.serviceClause.addServiceFlows(c.featureNetworkPolicy, []v1beta2.Service{udpService}, &dnsPriority, false)
-	dnsTCPMatchPairs := getServiceMatchPairs(tcpService, c.featureNetworkPolicy.ipProtocols)
-	for _, dnsTCPMatchPair := range dnsTCPMatchPairs {
-		tcpFlagsMatchPair := matchPair{
-			matchKey: MatchTCPFlags,
-			matchValue: TCPFlags{
-				// URG|ACK|PSH|RST|SYN|FIN|
-				Flag: 0b011000,
-				Mask: 0b011000,
-			},
-		}
-		if dnsTCPMatchPair[0].matchKey.GetOFProtocol() == binding.ProtocolTCPv6 {
-			tcpFlagsMatchPair.matchKey = MatchTCPv6Flags
-		}
+
+	tcpFlags := TCPFlags{
+		// URG|ACK|PSH|RST|SYN|FIN|
+		Flag: 0b011000,
+		Mask: 0b011000,
+	}
+	tcpDnsPort := types.BitRange{Value: uint16(dnsPort)}
+	for _, proto := range c.featureNetworkPolicy.ipProtocols {
 		tcpServiceMatch := &conjunctiveMatch{
-			tableID: conj.serviceClause.ruleTable.GetID(),
-			matchPairs: []matchPair{
```

Comment:
Thanks for finding the root cause of this issue. 
I think you can simply fix it by changing `matchPairs` here to `matchPairs: append(dnsTCPMatchPair, tcpFlagsMatchPair),`
Before adding source port support, `dnsTCPMatchPair` only has one item inside. Now, for a service with a source port match, `getServiceMatchPairs` will generate two pairs. One is for the destination port, matching port 0(match all) and the other is for the source port, matching the port we provided.
cc @Dyanngg since he is the owner of source port support.

Reply:
The truth is we only care about TCP source port and TCP flags, why it is a must to add a match for dst port value 0 although it does not work on OVS?

Reply:
Yes, from my understanding, that part could be improved. Let's see @Dyanngg 's opinion.